src/norm/mongodb.nim

```diff
@@ -0,0 +1,727 @@
+    result["thePet"] = toBson(obj.thePet, force)
+
+  ```
+  ]##
+  var
+    proc_map = initOrderedTable[string, string]() # object: procedure string
+    objectName = ""
+    typeName = ""
+    fieldName = ""
+    bsonFieldName = ""
+    key = ""
+
+  #
+  # generate one toBson per object
+  #
+  for obj in dbObjReprs:
+    objectName = obj.signature.name
+    key = objectName
+    proc_map[key] =  "proc toBson(obj: $1, force = false): Bson {.used.} =\n".format(objectName)
+    proc_map[key] &= "  result = newBsonDocument()\n"
+    proc_map[key] &= "\n"
+    #
+    # handle universal types first
+    #
+    for field in obj.fields:
+      typeName = reconstructType(field.typ)
+      fieldName = field.signature.name
+      bsonFieldName = fieldName
+      for p in field.signature.pragmas:
+        if p.name=="dbCol":
+          bsonFieldName = $p.value
+      if not NORM_UNIVERSAL_TYPE_LIST.contains(typeName):
+        continue
+      if typeName == "Oid":
+        proc_map[key] &= "  if $$obj.$1 != \"000000000000000000000000\":\n".format(fieldName)
+        proc_map[key] &= "    result[\"$1\"] = toBson(obj.$2)\n".format(bsonFieldName, fieldName)
+      elif typeName == "Time":
+        proc_map[key] &= "  if obj.$1 != fromUnix(0):\n".format(fieldName)
+        proc_map[key] &= "    result[\"$1\"] = toBson(obj.$2)\n".format(bsonFieldName, fieldName)
+      elif typeName.startsWith("Option["):
+        proc_map[key] &= "  if obj.$1.isNone:\n".format(fieldName)
+        proc_map[key] &= "    result[\"$1\"] = null()\n".format(bsonFieldName)
+        proc_map[key] &= "  else:\n"
+        proc_map[key] &= "    result[\"$1\"] = toBson(obj.$2.get())\n".format(bsonFieldName, fieldName)
+      elif typeName.startsWith("seq["):
+        proc_map[key] &= "  result[\"$1\"] = newBsonArray()\n".format(bsonFieldName)
+        proc_map[key] &= "  for entry in obj.$1:\n".format(fieldName)
+        proc_map[key] &= "    result[\"$1\"].add toBson(entry)\n".format(bsonFieldName)
+      else:
+        proc_map[key] &= "  result[\"$1\"] = toBson(obj.$2)\n".format(bsonFieldName, fieldName)
+    #
+    # now handle cross-object references
+    #
+    for field in obj.fields:
+      typeName = reconstructType(field.typ)
+      fieldName = field.signature.name
+      if NORM_UNIVERSAL_TYPE_LIST.contains(typeName):
+        continue
+      if normObjectNamesRegistry.contains(typeName):
+        proc_map[key] &= "  result[\"$1\"] = toBson(obj.$1, force)\n".format(fieldName)
+      else:
+        raise newException(
+          KeyError, 
+          "In object $1, the field $2's type is not known to norm/mongodb. If it is a subtending object, is $3 defined by dB, dbAddCollection, or dbAddObject yet?".format(objectName, fieldName, typeName)
+        )
+  #
+  # finish up all procedure strings
+  #
+  for key, s in proc_map.pairs():
+    result &= s
+    result &= "\n" # add a blank line between each proc
+
+const TYPE_TO_BSON_KIND = {

```

Comment:
Constants should live next to imports, at the very top.

Reply:
These constants only have meaning in the context of the procedure they precede. I'll move them if you really want it though.

Reply:
They have been moved to the top and are now compile-time variables.